src/analysis/refactor_extract_region.ml

```diff
@@ -0,0 +1,395 @@
+
+let () =
+  Location.register_error_of_exn (function
+    | Nothing_to_do -> Some (Location.error "Nothing to do")
+    | Not_allowed_in_interface_file ->
+      Some
+        (Location.error
+           "Expression extraction is only allowed in implementation file")
+    | _ -> None)
+
+module Fresh_name = struct
+  (* Generate a fresh name that does not already exist in given environment. *)
+  let gen_val_name basename env =
+    let rec loop n =
+      let guess = basename ^ Int.to_string n in
+      if Env.bound_value guess env then succ n |> loop else guess
+    in
+    loop 1
+end
+
+module Gen = struct
+  let unit = Longident.Lident "()" |> Location.mknoloc
+
+  (* Generates [let name = body]. *)
+  let toplevel_let ~name ~body =
+    let open Ast_helper in
+    let pattern = Pat.mk (Ppat_var { txt = name; loc = Location.none }) in
+    let body = Parsetree_utils.filter_expr_attr body in
+    Str.value Nonrecursive [ Vb.mk pattern body ]
+
+  (* Generates [let name () = body]. *)
+  let let_unit_toplevel ~name ~body =
+    let open Ast_helper in
+    let unit_param =
+      { Parsetree.pparam_loc = Location.none;
+        pparam_desc = Pparam_val (Nolabel, None, Pat.construct unit None)
+      }
+    in
+    let body = Exp.function_ [ unit_param ] None (Pfunction_body body) in
+    toplevel_let ~name ~body
+
+  (* Generates [let name params = body]. *)
+  let toplevel_function params ~name ~body =
+    let open Ast_helper in
+    let params =
+      List.map
+        ~f:(fun param ->
+          let pattern =
+            Pat.construct
+              (Location.mknoloc (Untypeast.lident_of_path param))
+              None
+          in
+          { Parsetree.pparam_loc = Location.none;
+            pparam_desc = Pparam_val (Nolabel, None, pattern)
+          })
+        params
+    in
+    let body = Exp.function_ params None (Pfunction_body body) in
+    toplevel_let ~name ~body
+
+  let ident ~name =
+    Longident.Lident name |> Location.mknoloc |> Ast_helper.Exp.ident
+
+  let fun_apply params ~name =
+    let open Ast_helper in
+    let params = List.map ~f:(fun p -> (Asttypes.Nolabel, p)) params in
+    Exp.apply (ident ~name) params
+
+  let fun_apply_unit = fun_apply [ Ast_helper.Exp.ident unit ]
+
+  let fun_apply_params params =
+    params
+    |> List.map ~f:(fun param ->
+           Ast_helper.Exp.ident
+             (Location.mknoloc (Longident.Lident (Path.name param))))
+    |> fun_apply
+end
+
+module Msource = struct
+  include Msource
+
+  (* TODO: Maybe add this directly in [Msource]? *)
+  let sub_loc src loc =
+    let (`Offset start_offset) =
+      let line, col = Lexing.split_pos loc.Location.loc_start in
+      Msource.get_offset src (`Logical (line, col))
+    in
+    let (`Offset end_offset) =
+      `Logical (Lexing.split_pos loc.loc_end) |> Msource.get_offset src
+    in
+    String.sub (Msource.text src) ~pos:start_offset
+      ~len:(end_offset - start_offset)
+    |> Msource.make
+end
```

Comment:
Not sure about the pattern `module A = struct include A, and reuse A as fully qualified expr`. 
You can have a function, outside of your `Msource`?

Reply:
Fixed.